File: Assistant/lib/utility/command_line_reader.dart
import '/common.dart';
import 'package:collection/collection.dart';

// ----------------

class CommandLineReader {

  // #region structor

  List<String> _view;

  Integer _position;

  // ----------------

  CommandLineReader(
    List<String> view,
  ) :
    _view = view,
    _position = 0;

  // #endregion

  // #region query

  Boolean done(
  ) {
    return !(0 <= this._position && this._position < this._view.length);
  }

  Boolean check(
    String name,
  ) {
    var state = !this.done() && this._view[this._position] == name;
    if (state) {
      this._position++;
    }
    return state;
  }

  // ----------------

  Boolean nextBoolean(
  ) {
    return this.nextString().selfLet((it) => Boolean.parse(it));
  }

  Integer nextInteger(
  ) {
    return this.nextString().selfLet((it) => Integer.parse(it));
  }

  Floater nextFloater(
  ) {
    return this.nextString().selfLet((it) => Floater.parse(it));
  }

  String nextString(
  ) {
    assertTest(!this.done());
    var value = this._view[this._position];
    this._position++;
    return value;
  }

  List<String> nextStringList(
  ) {
    var value = this._view.slice(this._position);
    this._position = this._view.length;
    return value;
  }

  // #endregion

}

File: AssistantPlus/Application/Utility/CommandLineReader.cs
#pragma warning disable 0,
// ReSharper disable

using AssistantPlus;

namespace AssistantPlus.Utility {

	public class CommandLineReader {

		#region structor

		private List<String> mView;

		private Size mPosition;

		// ----------------

		public CommandLineReader (
			List<String> view
		) {
			this.mView = view;
			this.mPosition = 0;
		}

		#endregion

		#region query

		public Boolean Done (
		) {
			return !(0 <= this.mPosition && this.mPosition < this.mView.Count);
		}

		public Boolean Check (
			String name
		) {
			var state = !this.Done() && this.mView[this.mPosition] == name;
			if (state) {
				this.mPosition++;
			}
			return state;
		}

		// ----------------

		public Boolean NextBoolean (
		) {
			return this.NextString().SelfLet((it) => (Boolean.Parse(it)));
		}

		public Integer NextInteger (
		) {
			return this.NextString().SelfLet((it) => (Integer.Parse(it)));
		}

		public Floater NextFloater (
		) {
			return this.NextString().SelfLet((it) => (Floater.Parse(it)));
		}

		public String NextString (
		) {
			GF.AssertTest(!this.Done());
			var value = this.mView[this.mPosition];
			this.mPosition++;
			return value;
		}

		public List<String> NextStringList (
		) {
			var value = this.mView[this.mPosition..];
			this.mPosition = this.mView.Count;
			return value;
		}

		#endregion

	}

}
